announcement/models/announcement.py

```diff
@@ -0,0 +1,128 @@
+        ],
+        default="specific_users",
+        required=True,
+    )
+    specific_user_ids = fields.Many2many(
+        comodel_name="res.users",
+        domain=[("share", "=", False)],
+        inverse="_inverse_specific_user_ids",
+    )
+    user_group_ids = fields.Many2many(comodel_name="res.groups")
+    allowed_user_ids = fields.Many2many(
+        comodel_name="res.users",
+        relation="announcement_res_users_allowed_rel",
+        compute="_compute_allowed_user_ids",
+        compute_sudo=True,
+        store=True,
+    )
+    allowed_users_count = fields.Integer(
+        compute="_compute_allowed_user_ids", compute_sudo=True, store=True,
+    )
+    read_announcement_count = fields.Integer(
+        compute="_compute_read_announcement_count", store=True,
+    )
+    notification_date = fields.Datetime()
+    notification_expiry_date = fields.Datetime()
+
+    def _inverse_specific_user_ids(self):
+        """Used to set users pendant announcements when they're set in the announcement
+        itself"""
+        for announcement in self:
+            for user in announcement.specific_user_ids.filtered(
+                lambda x: announcement
+                not in (x.read_announcement_ids + x.pendant_announcement_ids)
+            ):
+                user.pendant_announcement_ids |= announcement
+
+    @api.depends("specific_user_ids", "user_group_ids")
+    def _compute_allowed_user_ids(self):
+        self.allowed_user_ids = False
+        self.allowed_users_count = False
+        for announcement in self.filtered(
+            lambda x: x.announcement_type == "specific_users"
+        ):
+            announcement.allowed_user_ids = announcement.specific_user_ids
+            announcement.allowed_users_count = len(announcement.specific_user_ids)
+        for announcement in self.filtered(
+            lambda x: x.announcement_type == "user_group"
+        ):
+            announcement.allowed_user_ids = announcement.user_group_ids.users
+            announcement.allowed_users_count = len(announcement.user_group_ids.users)
+
+    def _compute_read_announcement_count(self):
+        logs = self.env["announcement.log"].read_group(
+            [("announcement_id", "in", self.ids)],
+            ["announcement_id"],
+            ["announcement_id"],
+        )
+        result = {
+            data["announcement_id"][0]: (data["announcement_id_count"]) for data in logs
+        }
+        for announcement in self:
+            announcement.read_announcement_count = result.get(announcement.id, 0)
+
+    @api.onchange("announcement_type")
+    def _onchange_announcement_type(self):
+        if self.announcement_type == "specific_users":
+            self.user_group_ids = False
+        elif self.announcement_type == "user_group":
+            self.specific_user_ids = False
+
+    def action_announcement_log(self):
+        """Return a set of fungible transient records to see altogether read logs
+        and pendant users"""
```

Comment:
Please try to keep docstrings short w/ 1 line + description (as for commits). See https://peps.python.org/pep-0257/#multi-line-docstrings